File: app/src/main/java/com/example/testproject/network/interceptor/ProgressResponseBody.java
package com.example.testproject.network.interceptor;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.example.testproject.network.CallBack;
import okhttp3.MediaType;
import okhttp3.ResponseBody;
import okio.*;

import java.io.File;
import java.io.IOException;

/**
 * Created by mike.li.
 * 进度
 */

public class ProgressResponseBody extends ResponseBody {
    private ResponseBody responseBody;
    private static CallBack<File> callBack;

    public static void setCallBack(CallBack<File> callBack) {
        ProgressResponseBody.callBack = callBack;
    }

    public ProgressResponseBody(ResponseBody responseBody) {
        this.responseBody = responseBody;
    }
    @Nullable
    @Override
    public MediaType contentType() {
        return responseBody.contentType();
    }

    @Override
    public long contentLength() {
        return responseBody.contentLength();
    }

    @Override
    public BufferedSource source() {
        return Okio.buffer(source(responseBody.source()));
    }

    private Source source(Source source) {
        return new ForwardingSource(source) {
            long readLength = 0L;
            @Override
            public long read(@NonNull Buffer sink, long byteCount) throws IOException {
                long bytesRead = super.read(sink, byteCount);
                readLength += bytesRead != -1 ? bytesRead : 0;
                if(callBack!=null)callBack.onProgress(readLength,responseBody.contentLength());
                return bytesRead;
            }
        };
    }
}

File: app/src/main/java/com/example/testproject/network/interceptor/ProgressResponseBody.kt
package com.example.testproject.network.interceptor

import com.example.testproject.network.CallBack
import okhttp3.MediaType
import okhttp3.ResponseBody
import okio.*

import java.io.File
import java.io.IOException

/**
 * Created by mike.li.
 * 进度
 */

class ProgressResponseBody(private val responseBody: ResponseBody) : ResponseBody() {
    override fun contentType(): MediaType? {
        return responseBody.contentType()
    }

    override fun contentLength(): Long {
        return responseBody.contentLength()
    }

    override fun source(): BufferedSource {
        return Okio.buffer(source(responseBody.source()))
    }

    private fun source(source: Source): Source {
        return object : ForwardingSource(source) {
            internal var readLength = 0L
            @Throws(IOException::class)
            override fun read(sink: Buffer, byteCount: Long): Long {
                val bytesRead = super.read(sink, byteCount)
                readLength += if (bytesRead != -1L) bytesRead else 0
                if (callBack != null) callBack!!.onProgress(readLength, responseBody.contentLength())
                return bytesRead
            }
        }
    }

    companion object {
        private var callBack: CallBack<File>? = null

        fun setCallBack(callBack: CallBack<File>) {
            ProgressResponseBody.callBack = callBack
        }
    }
}
